feat: disable path overlap validation flag

Fsero opened the pull request:
<!--- Provide a general summary of your changes in the Title above ---> <!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). ---> <!--- Please make sure you title is descriptive, it is used in the Release notes to let others know what it does --->

<!--- Why is this change required? What problem does it solve? --> <!--- If it fixes an open issue, please link to the issue here. -->

In https://github.com/kubernetes/ingress-nginx/issues/5651 there was a request to throw an error when there are two ingresses defining the same host and path, which was implemented as part of the validation webhook.

  Despite of this there are clear rules on the ingress controller that
describes what the controller would do in ![such situation (the oldest rule wins)](https://github.com/kubernetes/ingress-nginx/blob/main/docs/how-it-works.md?plain=1#L27)

  Some users are relying on this validation behaviour to prevent
misconfigurations, but there are use cases where allowing it, maybe temporarily, is helpful. Those use cases includes:

  - Splitting large ingresses objects in smaller ones https://github.com/kubernetes/ingress-nginx/issues/10820
  - Moving ingress objects between namespaces without downtime (like when you transfer an ingress from team A that owns namespace A to team B that owns namespace B) https://github.com/kubernetes/ingress-nginx/issues/10090

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

fixes #10820
fixes #10090

<!--- Please describe in detail how you tested your changes. --> <!--- Include details of your testing environment, and the tests you ran to --> <!--- see how your change affects other areas of the code, etc. -->

<!--- Go over all the following points, and put an `x` in all the boxes that apply. --> <!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [X] I have added unit and/or e2e tests to cover my changes.
- [X] All new and existing tests passed.

Change-Id: I9d4124d1c36876b06d63100cd10988eaf2f41db9

<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->
<!--- Please make sure you title is descriptive, it is used in the Release notes to let others know what it does ---> 

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.

longwuyuan commented:
From the K8S official resources, it will be useful to add a link  here that covers the specs and KEP for any ingress-controller, to deal with duplicate routing rule factors like identical hostname and identical path.

strongjz commented:
I'm not sure about adding footguns like this to the configuration. I'm not sure if it's in the checkoverlap function. Can you make sure this dumps out a warning in the admission logs? 

Can you add an e2e test about this with it enabled and disabled?

strongjz commented:
/kind feature

strongjz commented:
/triage accepted

Fsero commented:
> I'm not sure about adding footguns like this to the configuration.
> 
> With that said, we can add it but disable it as default, so the cluster admin has to choose to allow it consciously.
> 
> It may be in the check overlap function, but can you ensure this dumps out a warning in the admission logs?
> 
> Can you add an e2e test about this with it enabled and disabled?

Definitely, let me take a look and I'd do the e2e test

ailurarctos commented:
I don't think this fixes https://github.com/kubernetes/ingress-nginx/issues/10090 but it does complement the fix for it (https://github.com/kubernetes/ingress-nginx/pull/10091).

My understanding is that this MR will allow users who make use of overlapping paths (to perform zero-downtime migrations) to still take advantage of the validation webhook (which is great!). The issue https://github.com/kubernetes/ingress-nginx/issues/10090 is that the NGINX ingress controller will generate an invalid configuration in specific cases where overlapping paths with both prefix and exact types are used (the example in https://github.com/kubernetes/ingress-nginx/issues/10090#issue-1760116420).

Fsero commented:
@strongjz I've added e2e tests, didn't add disabled ones as disabled is the default behavior and some test are already checking for path overlap (but if you think is required I can add it obviously, just wanted to keep tests readable) also added the warning when enabled the flag.

Tell me what you think

strongjz commented:
/lgtm

/hold 

ill let @rikatz take a look and then we can get it merged into main.

Probably going to into 1.11.0, trying not to add features in patch release.

KTamas commented:
Hi, is there an update on this? We're running into this exact issue and merging this PR would solve it.

tao12345666333 commented:
> Moving ingress objects between namespaces without downtime 

I believe that this use case has practical significance.


Please resolve the conflicts.

Fsero commented:
done @tao12345666333

Fsero commented:
@strongjz @rikatz @cpanato is there something else I can do to merge this one?

Thanks a lot

cpanato commented:
lets wait a bit more on this, i will discuss that with @rikatz when we both have a time

tao12345666333 commented:
We can add it to the list for our next community meeting.

Fsero commented:
> We can add it to the list for our next community meeting. @strongjz

hey! any news on this? 

thanks so much

airhorns commented:
What's the hesitation on this? It seems super valuable to me -- happy to provide more feedback about why if that'd be helpful!

Fsero commented:
@strongjz @rikatz @cpanato sometime has passed and this one it's still open, do you need something else that I can do?

Thanks!

Fsero commented:
resurfacing this again @strongjz @rikatz @cpanato do you need me to do something else? do you want me to join the community meeting to explain I think this is useful (we are using this internally in our fork)

Thanks a lot

tao12345666333 commented:
Sorry for the long delay. 

As I replied before, I think this case has practical significance.
https://github.com/kubernetes/ingress-nginx/pull/10943#issuecomment-2028233491

@Fsero @airhorns You can join the community meeting next time or post details here.
We want to know more about the background and the value of this use case.

airhorns commented:
My use case is described here: https://github.com/kubernetes/ingress-nginx/issues/10820 -- in order to make that change, we needed to manually remove the validating webhook in k8s which felt quite risky, make the change, then add it back after removing the duplicates. I can confirm the actual controller was fine with the duplicate resources. What extra background on that use case are you looking for?

KTamas commented:
This would be so incredibly useful.

Fsero commented:
let's see if i can explain an use case, imagine you have a company with several teams deploying microservices, there is a good amount of traffic in the service and in the ingress, and due to an internal change you want to transfer the service from team foo to team bar.

Original microservice offered two endpoints /apple and /banana and team bar will get ownership of /banana while /apple will remain managed by foo.

you can setup this scenario in a local kind  (I actually created one with make dev-env) and try to apply the following manifest

```
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: foo
  name: foo
spec:
  finalizers:
  - kubernetes
---
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: bar
  name: bar
spec:
  finalizers:
  - kubernetes
---
kind: Pod
apiVersion: v1
metadata:
  name: banana-app
  namespace: foo
  labels:
    app: banana
spec:
  containers:
    - name: banana-app
      image: hashicorp/http-echo
      args:
        - "-text=banana"
---
kind: Service
apiVersion: v1
metadata:
  name: banana-service
  namespace: foo
spec:
  selector:
    app: banana
  ports:
    - port: 5678 # Default port for image
---
kind: Pod
apiVersion: v1
metadata:
  name: apple-app
  namespace: bar
  labels:
    app: apple
spec:
  containers:
    - name: apple-app
      image: hashicorp/http-echo
      args:
        - "-text=apple"
---

kind: Service
apiVersion: v1
metadata:
  name: apple-service
  namespace: bar
spec:
  selector:
    app: apple
  ports:
    - port: 5678 # Default port for image
    
---


apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: one-ingress
  namespace: foo
  annotations:
    ingress.kubernetes.io/rewrite-target: /
spec:
  ingressClassName: nginx
  rules:
  - host: foo.bar.io
    http:
      paths:
        - path: /apple
          pathType: Prefix
          backend:
            service:
               name: apple-service
               port: 
                 number: 5678
        - path: /banana
          pathType: Prefix
          backend:
            service:
                name: banana-service
                port:
                  number: 5678
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: another-ingress
  namespace: bar
  annotations:
    ingress.kubernetes.io/rewrite-target: /
spec:
  ingressClassName: nginx
  rules:
  - host: foo.bar.io
    http:
      paths:
        - path: /banana
          pathType: Prefix
          backend:
            service:
                name: banana-service
                port:
                  number: 5678
```


bar team members have read this part of the [documentation](https://github.com/kubernetes/ingress-nginx/blob/main/docs/how-it-works.md?plain=1#L27) of ingress nginx and thought that would be very neat to create an ingress and deploy the service for some local test and allow the original service to handle the traffic so when everything is ready foo team just need to edit one-ingress object and remove the /banana path and team bar will start receiving traffic with 0 downtime.

but when someone in bar tries to apply the manifest it finds the following error
```
ingress.networking.k8s.io/one-ingress unchanged
Error from server (BadRequest): error when creating "manifest.yaml": admission webhook "validate.nginx.ingress.kubernetes.io" denied the request: host "foo.bar.io" and path "/banana" is already defined in ingress foo/one-ingress
```

for most cases that might be a valid protection but when you want to achieve zero downtime traffic migrations between ingress objects is counterproductive.

This is the purpose of this PR give an opt-in flag that allows maintainers to bypass this and only this validation without disabling the rest of admission validations that are still useful.

cc @tao12345666333 @strongjz @rikatz
